sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -158,111 +178,86 @@ abstract class CaseWhenBase(
-  }
-}
-
-/**
- * CaseWhen expression used when code generation condition is satisfied.
- * OptimizeCodegen optimizer replaces CaseWhen into CaseWhenCodegen.
- *
- * @param branches seq of (branch condition, branch value)
- * @param elseValue optional value for the else branch
- */
-case class CaseWhenCodegen(
-    val branches: Seq[(Expression, Expression)],
-    val elseValue: Option[Expression] = None)
-  extends CaseWhenBase(branches, elseValue) with Serializable {
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    // Generate code that looks like:
-    //
-    // condA = ...
-    // if (condA) {
-    //   valueA
-    // } else {
-    //   condB = ...
-    //   if (condB) {
-    //     valueB
-    //   } else {
-    //     condC = ...
-    //     if (condC) {
-    //       valueC
-    //     } else {
-    //       elseValue
-    //     }
-    //   }
-    // }
+    // This variable represents whether the first successful condition is met or not.
+    // It is initialized to `false` and it is set to `true` when the first condition which
+    // evaluates to `true` is met and therefore is not needed to go on anymore on the computation
+    // of the following conditions.
+    val conditionMet = ctx.freshName("caseWhenConditionMet")
+    ctx.addMutableState(ctx.JAVA_BOOLEAN, ev.isNull)
+    ctx.addMutableState(ctx.javaType(dataType), ev.value)
     val cases = branches.map { case (condExpr, valueExpr) =>
       val cond = condExpr.genCode(ctx)
       val res = valueExpr.genCode(ctx)
       s"""
-        ${cond.code}
-        if (!${cond.isNull} && ${cond.value}) {
-          ${res.code}
-          ${ev.isNull} = ${res.isNull};
-          ${ev.value} = ${res.value};
+        if(!$conditionMet) {
+          ${cond.code}
+          if (!${cond.isNull} && ${cond.value}) {
+            ${res.code}
+            ${ev.isNull} = ${res.isNull};
+            ${ev.value} = ${res.value};
+            $conditionMet = true;
+          }
         }
       """
     }
 
-    var generatedCode = cases.mkString("", "\nelse {\n", "\nelse {\n")
-
-    elseValue.foreach { elseExpr =>
+    val elseCode = elseValue.map { elseExpr =>
       val res = elseExpr.genCode(ctx)
-      generatedCode +=
-        s"""
+      s"""
+        if(!$conditionMet) {
           ${res.code}
           ${ev.isNull} = ${res.isNull};
           ${ev.value} = ${res.value};
-        """
+        }
+      """
     }
 
-    generatedCode += "}\n" * cases.size
+    val allConditions = cases ++ elseCode
+
+    val code = if (ctx.INPUT_ROW == null || ctx.currentVars != null) {
+        allConditions.mkString("\n")
+      } else {
+        // This generates code like:
+        // do {
+        //   conditionMet = caseWhen_1(i);
+        //   if(conditionMet) {
+        //     continue;
+        //   }
+        //   conditionMet = caseWhen_2(i);
+        //   if(conditionMet) {
+        //     continue;
+        //   }
+        //   ...
+        // } while (false);
```

Comment:
Can we simplified to
```
do {
  if (caseWhen_1(i)) {
    continue;
  }
  if (caseWhen_2(i)) {
    continue;
  }
}
```

Reply:
No, because in this way we would not set `conditionMet`. This can cause an error, in the case we have so many functions that their invocation goes beyond the 64KB limit: this problem is fixed by #19480, by creating a method which calls the methods of that class, In this case, if we don't set properly `conditionMet`, we would have a bug in the implementation.

Reply:
I don't get it, are you saying
```
do {
  if (caseWhen_1(i)) {
    continue;
  }
  if (caseWhen_2(i)) {
    continue;
  }
}
```
is worse than
```
do {
  condition = caseWhen_1(i);
  if (condition) {
    continue;
  }
  condition = caseWhen_2(i);
  if (condition) {
    continue;
  }
}
```
?

Reply:
I am saying that the first one is wrong since it doesn't set `condition` (which can be returned at the end of the method as of #19480), since it may return wrong results.

Reply:
I'll try to explain with an example. If we have a lot of methods, this can exceed 64KB:
```
do {
  condition = caseWhen_1(i);
  if (condition) {
    continue;
  }
  condition = caseWhen_2(i);
  if (condition) {
    continue;
  }
  // a lot of other methods here
}
```
Thus, #19480 can split them into:
```
do {
  condition = bunchOf_caseWhen_1(i);
  if (condition) {
    continue;
  }
  condition = bunchOf_caseWhen_2(i);
  if (condition) {
    continue;
  }
  // maybe some other here
} while (false);
...
InnerClass1 {
  boolean bunchOf_caseWhen_1(InternalRow i) {
   boolean condition = false;
    do {
      condition = caseWhen_1(i);
      if (condition) {
        continue;
      }
      condition = caseWhen_2(i);
      if (condition) {
        continue;
      }
      // a lot of other methods here
    } while (false);
    return condition;
  }
  ...
}
```
in this case, the implementation you are suggesting can return a wrong value in `bunchOf_caseWhen_1` and this will affect the correctness of the code.

Reply:
ah i see, we have to set the `condition` if the code is inside a method.
```
do {
  if (bunchOf_caseWhen_1(i)) {
    continue;
  }
  if (bunchOf_caseWhen_2(i)) {
    continue;
  }
  // maybe some other here
} while (false);
...
InnerClass1 {
  boolean bunchOf_caseWhen_1(InternalRow i) {
    do {
      if (caseWhen_1(i)) {
        return true;
      }
      if (caseWhen_2(i)) {
        continue;
      }
      // a lot of other methods here
    } while (false);
    return false;
  }
  ...
}
```
This would be the optimal but it's too much complexity with only a little gain.

Reply:
this would require a refactoring of the `splitExpressions` method which is really not worth IMHO. I will leave this as it is now, while I am addressing your other comment, thanks.